File: app/src/main/java/com/vikingsen/cheesedemo/model/data/comment/CommentRemoteDataSource.java
package com.vikingsen.cheesedemo.model.data.comment;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import com.vikingsen.cheesedemo.model.webservice.CheeseService;
import com.vikingsen.cheesedemo.model.webservice.dto.CommentDto;
import com.vikingsen.cheesedemo.model.webservice.dto.CommentRequestDto;
import com.vikingsen.cheesedemo.model.webservice.dto.CommentResponse;
import com.vikingsen.cheesedemo.util.NetworkUtil;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Maybe;
import io.reactivex.Single;
import retrofit2.Response;
import timber.log.Timber;

@Singleton
class CommentRemoteDataSource {

    private final CheeseService cheeseService;
    private final NetworkUtil networkUtil;

    @Inject
    CommentRemoteDataSource(CheeseService cheeseService,
                            NetworkUtil networkUtil) {
        this.cheeseService = cheeseService;
        this.networkUtil = networkUtil;
    }

    Single<List<CommentDto>> getComments(long cheeseId) {
        return Single.create(emitter -> {
            try {
                if (networkUtil.isConnected()) {
                    Response<List<CommentDto>> response = cheeseService.getComments(cheeseId).execute();
                    if (response.isSuccessful()) {
                        emitter.onSuccess(response.body());
                        return;
                    } else {
                        Timber.e("Failed to load comments for cheese %d (%s) : %s", cheeseId, response.code(), response.message());
                    }
                } else {
                    Timber.w("Network not connected");
                }
            } catch (Exception e) {
                Timber.e(e, "Exception fetching comments for cheese %d", cheeseId);
            }
            emitter.onSuccess(Collections.emptyList());
        });
    }

    @WorkerThread
    @Nullable
    Maybe<List<CommentResponse>> syncComments(@NonNull List<CommentRequestDto> requestDtos) {
        return Maybe.create(emitter -> {
            try {
                Response<List<CommentResponse>> response = cheeseService.postComment(requestDtos).execute();
                if (response.isSuccessful()) {
                    emitter.onSuccess(response.body());
                    return;
                } else {
                    Timber.e("Failed to post %d comments (%s) : %s", requestDtos.size(), response.code(), response.message());
                }
            } catch (Exception e) {
                Timber.e(e, "Exception fetching %d comments", requestDtos.size());
            }
            emitter.onComplete();
        });
    }
}

File: app/src/main/java/com/vikingsen/cheesedemo/model/data/comment/CommentRemoteDataSource.kt
package com.vikingsen.cheesedemo.model.data.comment

import android.support.annotation.WorkerThread
import com.vikingsen.cheesedemo.model.webservice.ApiResponse
import com.vikingsen.cheesedemo.model.webservice.CheeseService
import com.vikingsen.cheesedemo.model.webservice.dto.CommentDto
import com.vikingsen.cheesedemo.model.webservice.dto.CommentRequestDto
import com.vikingsen.cheesedemo.model.webservice.dto.CommentResponse
import com.vikingsen.cheesedemo.util.NetworkDisconnectedException
import com.vikingsen.cheesedemo.util.NetworkUtil
import io.reactivex.Maybe
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CommentRemoteDataSource
@Inject constructor(private val cheeseService: CheeseService,
                    private val networkUtil: NetworkUtil) {

    fun getComments(cheeseId: Long): ApiResponse<List<CommentDto>> {
        try {
            if (networkUtil.isConnected) {
                return ApiResponse(cheeseService.getComments(cheeseId).execute())
            }
        } catch (e: Exception) {
            Timber.e(e, "Exception fetching cheese")
        }
        return ApiResponse(NetworkDisconnectedException())
    }

    @WorkerThread
    fun syncComments(requestDtos: List<CommentRequestDto>): Maybe<List<CommentResponse>>? {
        return Maybe.create<List<CommentResponse>> { emitter ->
            try {
                val response = cheeseService.postComment(requestDtos).execute()
                if (response.isSuccessful) {
                    response.body()?.let {
                        emitter.onSuccess(it)
                    }
                    return@create
                } else {
                    Timber.e("Failed to post %d comments (%s) : %s", requestDtos.size, response.code(), response.message())
                }
            } catch (e: Exception) {
                Timber.e(e, "Exception fetching %d comments", requestDtos.size)
            }

            emitter.onComplete()
        }
    }
}
